[InstCombine] Extend folding of aggregate construction to cases when source aggregates are partially available

weiguozhi opened the pull request:
Function foldAggregateConstructionIntoAggregateReuse can fold
  insertvalue(phi(extractvalue(src1), extractvalue(src2)))
into
  phi(src1, src2)
when we can find source aggregates in all predecessors.

This patch extends it to handle following case
  insertvalue(phi(extractvalue(src1), elm2))
into
  phi(src1, insertvalue(elm2))
with the condition that the predecessor without source aggregate has only one successor.

weiguozhi committed:
[InstCombine] Extend folding of aggregate construction to cases when source aggregates are partially available

Function foldAggregateConstructionIntoAggregateReuse can fold
  insertvalue(phi(extractvalue(src1), extractvalue(src2)))
into
  phi(src1, src2)
when we can find source aggregates in all predecessors.

This patch extends it to handle following case
  insertvalue(phi(extractvalue(src1), elm2))
into
  phi(src1, insertvalue(elm2))
with the condition that the predecessor without source aggregate has only one
successor.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -1117,9 +1118,35 @@ Instruction *InstCombinerImpl::foldAggregateConstructionIntoAggregateReuse(
+      continue;
+
+    BasicBlock *Pred = It.first;
+    Builder.SetInsertPoint(Pred, Pred->getTerminator()->getIterator());
```
```
; opt -passes=instcombine reduced.ll -S
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define { ptr, i64 } @"_ZN91_$LT$core..str..iter..SplitWhitespace$u20$as$u20$core..iter..traits..iterator..Iterator$GT$4next17h160f3816bb012b94E"({ ptr, i64 } %0, ptr %1, i1 %.not) {
  br i1 %.not, label %3, label %4

3:                                                ; preds = %2
  store ptr null, ptr %1, align 8
  br label %5

4:                                                ; preds = %2
  %.elt1 = extractvalue { ptr, i64 } %0, 0
  br label %5

5:                                                ; preds = %4, %3
  %6 = phi ptr [ %.elt1, %4 ], [ null, %3 ]
  %.elt2 = extractvalue { ptr, i64 } %0, 1
  %7 = insertvalue { ptr, i64 } zeroinitializer, ptr %6, 0
  %8 = insertvalue { ptr, i64 } %7, i64 %.elt2, 1
  ret { ptr, i64 } %8
}
```
```
dtcxzyw@dtcxzyw:~/WorkSpace/Projects/compilers/LLVM/llvm-build$ bin/opt -passes=instcombine -S reduced.ll
Instruction does not dominate all uses!
  %.elt2 = extractvalue { ptr, i64 } %0, 1
  %4 = insertvalue { ptr, i64 } { ptr null, i64 poison }, i64 %.elt2, 1
LLVM ERROR: Broken module found, compilation aborted!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: bin/opt -passes=instcombine -S reduced.ll
1.      Running pass "verify" on module "reduced.ll"
 #0 0x0000722e04e0f222 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.20.0git+0x20f222)
 #1 0x0000722e04e0c0ef llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.20.0git+0x20c0ef)
 #2 0x0000722e04e0c235 SignalHandler(int) Signals.cpp:0:0
 #3 0x0000722e04842520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x0000722e048969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x0000722e048969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x0000722e048969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x0000722e04842476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x0000722e048287f3 abort ./stdlib/abort.c:81:7
 #9 0x0000722e04c6c56b llvm::report_fatal_error(llvm::Twine const&, bool) (.cold) ErrorHandling.cpp:0:0
#10 0x0000722e04d1b1a0 unsigned long std::uniform_int_distribution<unsigned long>::operator()<std::random_device>(std::random_device&, std::uniform_int_distribution<unsigned long>::param_type const&) (.isra.0) ExponentialBackoff.cpp:0:0
#11 0x0000722dfed8f642 llvm::VerifierPass::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x38f642)
#12 0x0000722e04fff3c5 llvm::detail::PassModel<llvm::Module, llvm::VerifierPass, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.20.0git+0x1f3c5)
#13 0x0000722dfed4b11a llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x34b11a)
#14 0x0000722e0500d1d7 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.20.0git+0x2d1d7)
#15 0x0000722e05019192 optMain (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.20.0git+0x39192)
#16 0x0000722e04829d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#17 0x0000722e04829e40 call_init ./csu/../csu/libc-start.c:128:20
#18 0x0000722e04829e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#19 0x000063043daba095 _start (bin/opt+0x1095)
Aborted (core dumped)
```

weiguozhi replied:
It is because I expect DoPHITranslation translates a value into another value accessible in predecessor, if a value  is defined in successor, I expect DoPHITranslation returns null, but it actually returns the value itself, it is not accessible from predecessor, so it causes the error.

weiguozhi committed:
If a value is defined in UseBB, we can't use it in PredBB, so we can't
add an insertvalue to PredBB.

arsenm commented on llvm/test/Transforms/InstCombine/fold-aggregate-reconstruction.ll:
```diff
@@ -213,3 +213,40 @@ exit:
+;
+  br i1 %.not, label %3, label %4
+
+3:                                                ; preds = %2
```
Use named values in tests

weiguozhi replied:
done

dtcxzyw reviewed (changes requested):
This patch causes InstCombine to hang when building https://github.com/jqlang/jq:
```
; timeout 5s bin/opt -passes=instcombine reduced.ll -S
define { i64, ptr } @minmax_by({ i64, ptr } %0, i1 %1, i1 %2, { i64, ptr } %3) {
  %5 = extractvalue { i64, ptr } %0, 0
  %6 = extractvalue { i64, ptr } %0, 1
  br label %7

7:                                                ; preds = %14, %4
  %.sroa.01.0 = phi i64 [ %5, %4 ], [ %.sroa.01.1, %14 ]
  %.sroa.3.0 = phi ptr [ %6, %4 ], [ %.sroa.3.1, %14 ]
  br i1 %1, label %8, label %15

8:                                                ; preds = %7
  br i1 %2, label %9, label %12

9:                                                ; preds = %8
  %10 = extractvalue { i64, ptr } %3, 0
  %11 = extractvalue { i64, ptr } %3, 1
  br label %14

12:                                               ; preds = %8
  %13 = getelementptr { i64, ptr }, ptr null, i32 0, i32 0
  br label %14

14:                                               ; preds = %12, %9
  %.sroa.01.1 = phi i64 [ %10, %9 ], [ %.sroa.01.0, %12 ]
  %.sroa.3.1 = phi ptr [ %11, %9 ], [ %.sroa.3.0, %12 ]
  br label %7

15:                                               ; preds = %7
  %.fca.0.insert = insertvalue { i64, ptr } zeroinitializer, i64 %.sroa.01.0, 0
  %.fca.1.insert = insertvalue { i64, ptr } %.fca.0.insert, ptr %.sroa.3.0, 1
  ret { i64, ptr } %.fca.1.insert
}
```

weiguozhi commented:
> This patch causes InstCombine to hang when building https://github.com/jqlang/jq:
> 
> ```
> ; timeout 5s bin/opt -passes=instcombine reduced.ll -S
> define { i64, ptr } @minmax_by({ i64, ptr } %0, i1 %1, i1 %2, { i64, ptr } %3) {
>   %5 = extractvalue { i64, ptr } %0, 0
>   %6 = extractvalue { i64, ptr } %0, 1
>   br label %7
> 
> 7:                                                ; preds = %14, %4
>   %.sroa.01.0 = phi i64 [ %5, %4 ], [ %.sroa.01.1, %14 ]
>   %.sroa.3.0 = phi ptr [ %6, %4 ], [ %.sroa.3.1, %14 ]
>   br i1 %1, label %8, label %15
> 
> 8:                                                ; preds = %7
>   br i1 %2, label %9, label %12
> 
> 9:                                                ; preds = %8
>   %10 = extractvalue { i64, ptr } %3, 0
>   %11 = extractvalue { i64, ptr } %3, 1
>   br label %14
> 
> 12:                                               ; preds = %8
>   %13 = getelementptr { i64, ptr }, ptr null, i32 0, i32 0
>   br label %14
> 
> 14:                                               ; preds = %12, %9
>   %.sroa.01.1 = phi i64 [ %10, %9 ], [ %.sroa.01.0, %12 ]
>   %.sroa.3.1 = phi ptr [ %11, %9 ], [ %.sroa.3.0, %12 ]
>   br label %7
> 
> 15:                                               ; preds = %7
>   %.fca.0.insert = insertvalue { i64, ptr } zeroinitializer, i64 %.sroa.01.0, 0
>   %.fca.1.insert = insertvalue { i64, ptr } %.fca.0.insert, ptr %.sroa.3.0, 1
>   ret { i64, ptr } %.fca.1.insert
> }
> ```

weiguozhi commented:
Thanks for the quick test.

The new insertvalue is added to a predecessor with only one successor, so it intends to reduce the number of executed insertvalue instruction. But it doesn't behave like this when a loop is involved. The loop exiting edges and loop latches don't guarantee reduced dynamic number of insertvalue instructions. We should check these cases.

weiguozhi committed:
Prevent add insertvalue to inner loops.

weiguozhi commented:
ping

dtcxzyw reviewed (changes requested):
Another reproducer causing instcombine to hang:
```
bin/opt -passes=instcombine reduced.ll
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

define { i64, ptr } @minmax_by({ i64, ptr } %0, i1 %1, i1 %2, { i64, ptr } %3) {
  %5 = extractvalue { i64, ptr } %0, 0
  %6 = extractvalue { i64, ptr } %0, 1
  br label %7

7:                                                ; preds = %14, %4
  %.sroa.01.0 = phi i64 [ %5, %4 ], [ %.sroa.01.1, %14 ]
  %.sroa.3.0 = phi ptr [ %6, %4 ], [ %.sroa.3.1, %14 ]
  br i1 %1, label %8, label %15

8:                                                ; preds = %7
  br i1 %2, label %9, label %12

9:                                                ; preds = %8
  %10 = extractvalue { i64, ptr } %3, 0
  %11 = extractvalue { i64, ptr } %3, 1
  br label %14

12:                                               ; preds = %8
  %13 = getelementptr { i64, ptr }, ptr null, i32 0, i32 0
  br label %14

14:                                               ; preds = %12, %9
  %.sroa.01.1 = phi i64 [ %10, %9 ], [ %.sroa.01.0, %12 ]
  %.sroa.3.1 = phi ptr [ %11, %9 ], [ %.sroa.3.0, %12 ]
  br label %7

15:                                               ; preds = %7
  %.fca.0.insert = insertvalue { i64, ptr } zeroinitializer, i64 %.sroa.01.0, 0
  %.fca.1.insert = insertvalue { i64, ptr } %.fca.0.insert, ptr %.sroa.3.0, 1
  ret { i64, ptr } %.fca.1.insert
}
```

BTW, I don't think this transform is beneficial since it blocks some select-related optimizations:
Before this patch:
```
define hidden { i64, i64 } @"_ZN4core6result19Result$LT$T$C$E$GT$7map_err17h84015741f1851455E.llvm.9397651224479784633"(i64 noundef %0, i64 %1) unnamed_addr #0 {
   %3 = insertvalue { i64, i64 } poison, i64 %0, 0
   %4 = insertvalue { i64, i64 } %3, i64 %1, 1
   ret { i64, i64 } %4
 }
```
After this patch:
```
define hidden { i64, i64 } @"_ZN4core6result19Result$LT$T$C$E$GT$7map_err17h84015741f1851455E.llvm.9397651224479784633"(i64 noundef %0, i64 %1) unnamed_addr #0 {
   %3 = icmp eq i64 %0, -9223372036854775807
   %4 = insertvalue { i64, i64 } poison, i64 %0, 0
   %5 = insertvalue { i64, i64 } %4, i64 %1, 1
   %.merged = select i1 %3, { i64, i64 } { i64 -9223372036854775807, i64 undef }, { i64, i64 } %5
   ret { i64, i64 } %.merged
 }
```

Can you provide some performance data to demonstrate the usefulness of this patch?

weiguozhi commented:
The test case is extracted from hot path of tcmalloc, so it has real world impact.

This patch uses LoopInfo, so you need to run it with
```
bin/opt -passes='instcombine<use-loop-info>' reduced.ll
```

I ran the patch on your smaller test case, and got
```
opt -passes='instcombine<use-loop-info>' reduced2.ll -S

define hidden { i64, i64 } @"_ZN4core6result19Result$LT$T$C$E$GT$7map_err17h84015741f1851E.llvm"(i64 noundef %0, i64 %1) unnamed_addr {
  %3 = insertvalue { i64, i64 } poison, i64 %0, 0
  %4 = insertvalue { i64, i64 } %3, i64 %1, 1
  ret { i64, i64 } %4
}
```
How did you run it?

dtcxzyw commented:
> The test case is extracted from hot path of tcmalloc, so it has real world impact.
> 
> This patch uses LoopInfo, so you need to run it with
> 
> ```
> bin/opt -passes='instcombine<use-loop-info>' reduced.ll
> ```
> 
> I ran the patch on your smaller test case, and got
> 
> ```
> opt -passes='instcombine<use-loop-info>' reduced2.ll -S
> 
> define hidden { i64, i64 } @"_ZN4core6result19Result$LT$T$C$E$GT$7map_err17h84015741f1851E.llvm"(i64 noundef %0, i64 %1) unnamed_addr {
>   %3 = insertvalue { i64, i64 } poison, i64 %0, 0
>   %4 = insertvalue { i64, i64 } %3, i64 %1, 1
>   ret { i64, i64 } %4
> }
> ```
> 
> How did you run it?

I run it with `opt -O3`. `UseLoopInfo` is disabled by default due to compilation time regression: https://github.com/llvm/llvm-project/commit/ef6f23535d4e2b437913b48bdbdef00f000594c2

weiguozhi commented:
> > How did you run it?
> 
> I run it with `opt -O3`. `UseLoopInfo` is disabled by default due to compilation time regression: [ef6f235](https://github.com/llvm/llvm-project/commit/ef6f23535d4e2b437913b48bdbdef00f000594c2)

With -O3 I got
```
opt -O3 reduced2.ll -S

; ModuleID = 'reduced2.ll'
source_filename = "reduced2.ll"

; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
define hidden { i64, i64 } @"_ZN4core6result19Result$LT$T$C$E$GT$7map_err17h84015741f1851E.llvm"(i64 noundef %0, i64 %1) unnamed_addr #0 {
  %3 = insertvalue { i64, i64 } poison, i64 %0, 0
  %4 = insertvalue { i64, i64 } %3, i64 %1, 1
  ret { i64, i64 } %4
}
```

Thank you for the LoopInfo information. In order to make it more usable I  should avoid the LoopInfo usage. I will think it more.

weiguozhi committed:
Use control flow and profile information to avoid infinite loop.

weiguozhi commented:
Now if I need to add insertvalue into predecessor, I restrict it to

1, UseBB == OrigBB, then it can not cross a loop
2, If profile is available, we can only add insertvalue into colder BB.

weiguozhi commented:
ping

nikic commented:
Some quick thoughts:
 * Looking at https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1288/files, a significant number of the transforms look non-profitable. This transforms is reasonable if it actually drops extractelement + insertelement pairs, but if you just move insertelements to a predecessor while converting scalar phis into aggregate phis, I think that's a loss, because optimizations generally have troubles with aggregate SSA values. I think this happens mostly if the other phi operands are constants.
 * BFI in InstCombine -- wat? It looks like it's currently only used in SLC. Let's not use it in InstCombine proper.
 * To protect against cycles you probably want to check isBackEdge() instead, which was recently added.
 * `succ_size(Pred) != 1` => This should probably explicitly check for an unconditional branch, so you don't need to deal with the weird edge cases? (Like say, a single-successor callbr with aggregate return).

weiguozhi committed:
Remove the usage of BFI

BFI in InstCombine has limited usage, it's usually not available, so
remove it. Also avoid constructing constant aggregate.

weiguozhi commented:
> * Looking at https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1288/files, a significant number of the transforms look non-profitable. This transforms is reasonable if it actually drops extractelement + insertelement pairs, but if you just move insertelements to a predecessor while converting scalar phis into aggregate phis, I think that's a loss, because optimizations generally have troubles with aggregate SSA values. I think this happens mostly if the other phi operands are constants.

The original code removes fully redundant insertvalue, this patch enhances it to remove partial redundant insertvalue, so it actually drop redundant insertvalue. Avoiding constant aggregate sounds reasonable, I added code to check it.

> * BFI in InstCombine -- wat? It looks like it's currently only used in SLC. Let's not use it in InstCombine proper.

Deleted.
> * To protect against cycles you probably want to check isBackEdge() instead, which was recently added.

Thanks for the info. Currently I restrict it to UseBB == OrigBB && succ_size(Pred) == 1, it can already avoid cycles.

> * `succ_size(Pred) != 1` => This should probably explicitly check for an unconditional branch, so you don't need to deal with the weird edge cases? (Like say, a single-successor callbr with aggregate return).

Changed it to checking for unconditional branch.

weiguozhi commented:
@dtcxzyw could you help to explain these numbers? Execution time? File size? Number of llvm IR instructions?
 
```
Top 5 improvements:
  minetest/s_client.cpp.ll 300581269 294367598 -2.07%
  abseil-cpp/kernel_timeout_test.cc.ll 1973629781 1937840415 -1.81%
  cmake/cmFindBase.cxx.ll 1913629020 1891008778 -1.18%
  assimp/IFCReaderGen2_2x3.cpp.ll 3046110312 3017099134 -0.95%
  darktable/lut3dgmic.cpp.ll 236183674 234289794 -0.80%
Top 5 regressions:
  minetest/test_connection.cpp.ll 682795465 698591227 +2.31%
  cmake/cmQtAutoMocUic.cxx.ll 8264379610 8383878253 +1.45%
  tokio-rs/132n8ebvfaa2dg8b.ll 82150296 82583502 +0.53%
  tokio-rs/40sgumesnmyyffj3.ll 67652318 67979824 +0.48%
  spike/f128_rem.ll 123368228 123914940 +0.44%

Overall: 0.01762297%
```

I can guess out the numbers in other two parts, the llvm stats result and the changed lines in diff result.

dtcxzyw commented:
> @dtcxzyw could you help to explain these numbers? Execution time? File size? Number of llvm IR instructions?
> 
> ```
> Top 5 improvements:
>   minetest/s_client.cpp.ll 300581269 294367598 -2.07%
>   abseil-cpp/kernel_timeout_test.cc.ll 1973629781 1937840415 -1.81%
>   cmake/cmFindBase.cxx.ll 1913629020 1891008778 -1.18%
>   assimp/IFCReaderGen2_2x3.cpp.ll 3046110312 3017099134 -0.95%
>   darktable/lut3dgmic.cpp.ll 236183674 234289794 -0.80%
> Top 5 regressions:
>   minetest/test_connection.cpp.ll 682795465 698591227 +2.31%
>   cmake/cmQtAutoMocUic.cxx.ll 8264379610 8383878253 +1.45%
>   tokio-rs/132n8ebvfaa2dg8b.ll 82150296 82583502 +0.53%
>   tokio-rs/40sgumesnmyyffj3.ll 67652318 67979824 +0.48%
>   spike/f128_rem.ll 123368228 123914940 +0.44%
> 
> Overall: 0.01762297%
> ```
> 
> I can guess out the numbers in other two parts, the llvm stats result and the changed lines in diff result.

It is the number of instructions executed by `opt -O3`.

weiguozhi commented:
> It is the number of instructions executed by `opt -O3`.

So it's the number of instructions executed by opt, not the number of instructions executed by the generated code. Am I right?

arsenm commented:
> > It is the number of instructions executed by `opt -O3`.
> 
> So it's the number of instructions executed by opt, not the number of instructions executed by the generated code. Am I right?

Yes

weiguozhi commented:
> > > It is the number of instructions executed by `opt -O3`.
> > 
> > 
> > So it's the number of instructions executed by opt, not the number of instructions executed by the generated code. Am I right?
> 
> Yes

Thanks for the confirmation. So it seems compile time impact is neutral. 

I can only see two files from https://github.com/dtcxzyw/llvm-opt-benchmark/commit/78a3cecad7713c6678e2f2d5e6c888f26b4d6b51. Both files reduced the dynamic number of insertvalue instructions, so it has positive performance impact.

weiguozhi commented:
ping

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -1117,9 +1125,68 @@ Instruction *InstCombinerImpl::foldAggregateConstructionIntoAggregateReuse(
+      // If UseBB is the single successor of Pred, we can add InsertValue to
+      // Pred.
+      auto *BI = dyn_cast<BranchInst>(Pred->getTerminator());
+      if (!(BI && BI->isUnconditional()))
```
```suggestion
      if (!BI || !BI->isUnconditional())
```

weiguozhi replied:
done

weiguozhi committed:
Change the condition expression.

weiguozhi commented:
ping

dtcxzyw commented:
I don't think we can benefit from introducing more aggregate phi nodes. If these phi nodes are converted into selects in subsequent passes (e.g., SimplifyCFG), it is hard for InstCombine to fold. See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1351#discussion_r1802502189.

If you think this patch will reduce total number of instructions, I don't agree with this argument. Can you provide some performance data on SPEC/llvm-test-suite?

Compilation time impact looks ok. Don't worry about this :)

I will accept this patch if:
1. most of the regressions related to `selects on aggregate types` are fixed.
2. This patch results in a net performance improvement on some benchmarks.

weiguozhi commented:
Add new instcombine pattern to transform
```
         extract (select (cond, insert(agg, elem), FV))
     ->  select (cond, elem, extract(FV))
```
so the regression of https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1351#discussion_r1802502189 can be fixed.

weiguozhi commented:
I tested it on speccpu 2006 int, this patch is triggerred 19 times for 483.xalancbmk only, but the generated code is same. So no impact.

The real world impact of this patch is to a hot path in tcmalloc. In function _ZN8tcmalloc17tcmalloc_internal14TCMallocPolicyINS0_12CppOomPolicyENS0_18DefaultAlignPolicyENS0_25AllocationAccessHotPolicyENS0_17InvokeHooksPolicyENS0_21IsSizeReturningPolicyENS0_24LocalNumaPartitionPolicyEE10to_pointerEPvmocPolicyINS0_12CppOomPolicyENS0_18DefaultAlignPolicyENS0_25AllocationAccessHotPolicyENS0_17InvokeHooksPolicyENS0_21IsSizeReturningPolicyENS0_24LocalNumaPartitionPolicyEE10to_pointerEPvm (will be inlined into tcmalloc_size_returning_operator_new) there is a code sequence similar to
```
BB1:
   {ptr, i64} %v1 = call {ptr, i64} @func1
   %f1 = extractvalue { ptr, i64 } %v1, 0
   %f2 = extractvalue { ptr, i64 } %v1, 1
   br label %exit

BB2:
   {ptr, i64} %v2 = call {ptr, i64} @_ZN8tcmalloc17tcmalloc_internal14TCMallocPolicyINS0_12CppOomPolicyENS0_18DefaultAlignPolicyENS0_25AllocationAccessHotPolicyENS0_17InvokeHooksPolicyENS0_21IsSizeReturningPolicyENS0_24LocalNumaPartitionPolicyEE10to_pointerEPvmocPolicyINS0_12CppOomPolicyENS0_18DefaultAlignPolicyENS0_25AllocationAccessHotPolicyENS0_17InvokeHooksPolicyENS0_21IsSizeReturningPolicyENS0_24LocalNumaPartitionPolicyEE10to_pointerEPvm 
   %f3 = extractvalue { ptr, i64 } %v2, 0
   %f4 = extractvalue { ptr, i64 } %v2, 1
   br label %exit

exit:
   %f5 = phi ptr [%f1, %BB1], [%f3, %BB2]
   %f6 = phi i64 [%f2, %BB1], [%f4, %BB2]
   %res0 = insertvalue { ptr, i64 } poison, ptr %f5, 0
   %res = insertvalue { ptr, i64 } %res0, i64 %f6, 1
   ret {ptr, i64} %res
```
Usually function foldAggregateConstructionIntoAggregateReuse can optimize away all of the insertvalue/extractvalue,
```
BB1:
   {ptr, i64} %v1 = call {ptr, i64} @func1
   br label %exit

BB2:
   {ptr, i64} %v2 = call {ptr, i64} @_ZN8tcmalloc17tcmalloc_internal14TCMallocPolicyINS0_12CppOomPolicyENS0_18DefaultAlignPolicyENS0_25AllocationAccessHotPolicyENS0_17InvokeHooksPolicyENS0_21IsSizeReturningPolicyENS0_24LocalNumaPartitionPolicyEE10to_pointerEPvmocPolicyINS0_12CppOomPolicyENS0_18DefaultAlignPolicyENS0_25AllocationAccessHotPolicyENS0_17InvokeHooksPolicyENS0_21IsSizeReturningPolicyENS0_24LocalNumaPartitionPolicyEE10to_pointerEPvm 
   br label %exit

exit:
   %res = phi {ptr, i64} [%v1, BB1], [%v2, BB2]
   ret {ptr, i64} %res
```
But in one of our applications built with thinlto, with profile information thinlto causes the function _ZN8tcmalloc17tcmalloc_internal14TCMallocPolicyINS0_12CppOomPolicyENS0_18DefaultAlignPolicyENS0_25AllocationAccessHotPolicyENS0_17InvokeHooksPolicyENS0_21IsSizeReturningPolicyENS0_24LocalNumaPartitionPolicyEE10to_pointerEPvmocPolicyINS0_12CppOomPolicyENS0_18DefaultAlignPolicyENS0_25AllocationAccessHotPolicyENS0_17InvokeHooksPolicyENS0_21IsSizeReturningPolicyENS0_24LocalNumaPartitionPolicyEE10to_pointerEPvm to be inlined, so I got 
```
BB1:
   {ptr, i64} %v1 = call {ptr, i64} @func1
   %f1 = extractvalue { ptr, i64 } %v1, 0
   %f2 = extractvalue { ptr, i64 } %v1, 1
   br label %exit

BB2:
   %f3 = phi ptr [], []
   %f4 = load i64 ...
   br label %exit

exit:
   %f5 = phi ptr [%f1, %BB1], [%f3, %BB2]
   %f6 = phi i64 [%f2, %BB1], [%f4, %BB2]
   %res0 = insertvalue { ptr, i64 } poison, ptr %f5, 0
   %res = insertvalue { ptr, i64 } %res0, i64 %f6, 1
   ret {ptr, i64} %res
```
This instruction sequence can't be handled by the original foldAggregateConstructionIntoAggregateReuse, so worse tcmalloc_size_returning_operator_new was generated.

dtcxzyw commented on llvm/test/Transforms/InstCombine/fold-aggregate-reconstruction.ll:
```diff
@@ -0,0 +1,353 @@
+  %tmp = insertvalue { ptr, i64 } poison, ptr %val1, 0
+  %res = insertvalue { ptr, i64 } %tmp, i64 %val2, 1
+  ret {ptr, i64} %res
+}
```
But this optimization doesn't improve the final codegen: https://godbolt.org/z/YrW8zvo3x

Does it improve tcmalloc's performance?

weiguozhi replied:
Add "tail" to the call instruction, then you can see the function call is changed to a tailcall, and function prolog and epilog is simplified(removed). In the real code the %exit block has 4 or 5 predecessors ending with tail call.

In our application it does improve tcmalloc's performance because tcmalloc_size_returning_operator_new is a hot function.

dtcxzyw commented:
IR diff looks good to me. I believe the regression can be addressed by using `simplifyInstruction` in `FoldOpIntoSelect`: https://alive2.llvm.org/ce/z/HiwgPk. We discussed about this in https://github.com/llvm/llvm-project/pull/84686#issuecomment-1987344128.

weiguozhi commented:
ping

dtcxzyw commented:
> IR diff looks good to me. I believe the regression can be addressed by using `simplifyInstruction` in `FoldOpIntoSelect`: https://alive2.llvm.org/ce/z/HiwgPk. We discussed about this in [#84686 (comment)](https://github.com/llvm/llvm-project/pull/84686#issuecomment-1987344128).

@weiguozhi Are you willing to try this?

weiguozhi committed:
Change the loop iterator.

weiguozhi commented:
> > IR diff looks good to me. I believe the regression can be addressed by using `simplifyInstruction` in `FoldOpIntoSelect`: https://alive2.llvm.org/ce/z/HiwgPk. We discussed about this in [#84686 (comment)](https://github.com/llvm/llvm-project/pull/84686#issuecomment-1987344128).
> 
> @weiguozhi Are you willing to try this?

Yes, I tried it. As you have noticed it doesn't handle non-const aggregates. We can enhance FoldOpIntoSelect to do the non-const transform in https://github.com/llvm/llvm-project/pull/100828/commits/aef54025298b496d4f0b8766816700659b2f9015

         extract (select (cond, insert(agg, elem), FV))
     ->  select (cond, elem, extract(FV))

dtcxzyw commented:
> > > IR diff looks good to me. I believe the regression can be addressed by using `simplifyInstruction` in `FoldOpIntoSelect`: https://alive2.llvm.org/ce/z/HiwgPk. We discussed about this in [#84686 (comment)](https://github.com/llvm/llvm-project/pull/84686#issuecomment-1987344128).
> > 
> > 
> > @weiguozhi Are you willing to try this?
> 
> Yes, I tried it. As you have noticed it doesn't handle non-const aggregates. We can enhance FoldOpIntoSelect to do the non-const transform in [aef5402](https://github.com/llvm/llvm-project/commit/aef54025298b496d4f0b8766816700659b2f9015)
> 
> ```
>      extract (select (cond, insert(agg, elem), FV))
>  ->  select (cond, elem, extract(FV))
> ```

Can you file a separate PR?

weiguozhi committed:
Revert "Do the transform"

This reverts commit aef54025298b496d4f0b8766816700659b2f9015.
This optimization has been implemented in
https://github.com/llvm/llvm-project/commit/9a844a36eb9a21de27882b6193a82fda49986347.

weiguozhi committed:
Rebase test case.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -1117,9 +1124,68 @@ Instruction *InstCombinerImpl::foldAggregateConstructionIntoAggregateReuse(
+    // Avoid constructing constant aggregate because constant value may expose
+    // more optimizations.
+    bool ConstAgg = true;
+    for (auto [Idx, Val] : enumerate(AggElts)) {
```
```suggestion
    for (auto Val : AggElts) {
```
`Idx` is unused.

weiguozhi replied:
done.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -1117,9 +1124,68 @@ Instruction *InstCombinerImpl::foldAggregateConstructionIntoAggregateReuse(
+      continue;
+
+    BasicBlock *Pred = It.first;
+    Builder.SetInsertPoint(Pred, Pred->getTerminator()->getIterator());
```
```suggestion
    Builder.SetInsertPoint(Pred->getTerminator());
```

weiguozhi replied:
done.

dtcxzyw reviewed (approved):
LGTM. Please wait for additional approval from other reviewers.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -1117,9 +1124,68 @@ Instruction *InstCombinerImpl::foldAggregateConstructionIntoAggregateReuse(
+  }
+
+  if (!FoundSrcAgg)
+    return nullptr;
```
If I'm understanding correctly, this currently checks that at least one predecessor folds, rather than that all but one folds, so we could end up inserting extra insertvalues in *many* predecessors. Shouldn't we be limiting this such that we only insert into at most one predecessor?

nikic replied:
For example, here is a slight variant of test2 with the current patch:
```
define {ptr, i64} @test2(i1 %cond1, i1 %cond2, ptr %p1, ptr %p2) {
; CHECK-LABEL: define { ptr, i64 } @test2(
; CHECK-SAME: i1 [[COND1:%.*]], i1 [[COND2:%.*]], ptr [[P1:%.*]], ptr [[P2:%.*]]) {
; CHECK-NEXT:    br i1 [[COND1]], label %[[BBB1:.*]], label %[[BBB4:.*]]
; CHECK:       [[BBB1]]:
; CHECK-NEXT:    br i1 [[COND2]], label %[[BBB2:.*]], label %[[BBB3:.*]]
; CHECK:       [[BBB2]]:
; CHECK-NEXT:    [[CALL1:%.*]] = call { ptr, i64 } @bar(i64 0)
; CHECK-NEXT:    br label %[[EXIT:.*]]
; CHECK:       [[BBB3]]:
; CHECK-NEXT:    [[VAL31:%.*]] = load ptr, ptr [[P1]], align 8
; CHECK-NEXT:    [[VAL32:%.*]] = load i64, ptr [[P2]], align 4
; CHECK-NEXT:    [[TMP1:%.*]] = insertvalue { ptr, i64 } poison, ptr [[VAL31]], 0
; CHECK-NEXT:    [[TMP2:%.*]] = insertvalue { ptr, i64 } [[TMP1]], i64 [[VAL32]], 1
; CHECK-NEXT:    br label %[[EXIT]]
; CHECK:       [[BBB4]]:
; CHECK-NEXT:    [[VAL33:%.*]] = load ptr, ptr [[P1]], align 8
; CHECK-NEXT:    [[VAL34:%.*]] = load i64, ptr [[P2]], align 4
; CHECK-NEXT:    [[TMP3:%.*]] = insertvalue { ptr, i64 } poison, ptr [[VAL33]], 0
; CHECK-NEXT:    [[TMP4:%.*]] = insertvalue { ptr, i64 } [[TMP3]], i64 [[VAL34]], 1
; CHECK-NEXT:    br label %[[EXIT]]
; CHECK:       [[EXIT]]:
; CHECK-NEXT:    [[RES_MERGED:%.*]] = phi { ptr, i64 } [ [[CALL1]], %[[BBB2]] ], [ [[TMP2]], %[[BBB3]] ], [ [[TMP4]], %[[BBB4]] ]
; CHECK-NEXT:    ret { ptr, i64 } [[RES_MERGED]]
;
  br i1 %cond1, label %bbb1, label %bbb4

bbb1:
  br i1 %cond2, label %bbb2, label %bbb3

bbb2:
  %call1 = call {ptr, i64} @bar(i64 0)
  %val11 = extractvalue { ptr, i64 } %call1, 0
  %val12 = extractvalue { ptr, i64 } %call1, 1
  br label %exit

bbb3:
  %val21 = load ptr, ptr %p1
  %val22 = load i64, ptr %p2
  br label %exit

bbb4:
  %val31 = load ptr, ptr %p1
  %val32 = load i64, ptr %p2
  br label %exit

exit:
  %val1 = phi ptr [%val11, %bbb2], [%val21, %bbb3], [%val31, %bbb4]
  %val2 = phi i64 [%val12, %bbb2], [%val22, %bbb3], [%val32, %bbb4]
  %tmp = insertvalue { ptr, i64 } poison, ptr %val1, 0
  %res = insertvalue { ptr, i64 } %tmp, i64 %val2, 1
  ret {ptr, i64} %res
}
```

We remove extractvalues in one branch, but trade this with insertvalues in two branches. This doesn't look profitable, especially once we extend this to even more branches that don't fold.

dtcxzyw replied:
> This doesn't look profitable, especially once we extend this to even more branches that don't fold.

The latter one is better: https://godbolt.org/z/Yfhjhd3Yn

weiguozhi replied:
Because we limits the insertvalue to be moved to immediate predecessors and UseBB is the only successor of PredBB, the total execution count of UseBB is the sum of all predecessors. So once we have found some source aggregate in a PredBB, add insertvalue to all other predecessors still results in less dynamic count of insertvalue instructions.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -1117,9 +1124,68 @@ Instruction *InstCombinerImpl::foldAggregateConstructionIntoAggregateReuse(
+    // If OrigIVI is in UseBB and it's the only successor of PredBB, PredBB
+    // can't be in inner loop.
+    if (UseBB != OrigBB)
+      return nullptr;
```
I have a hard time understanding why this check is sufficient to prevent cycles in all cases, or even how the UseBB != OrigBB condition is related. I guess this is supposed to check the case where OrigBB is a loop header with phis and UseBB is an exit block with inserts -- but couldn't it fail on non-sunk inserts?

weiguozhi replied:
It limits the insertvalue can only be inserted into immediate predecessors. When combining the previous condition that UseBB is the only successor of PredBB, the insertvalue can never be moved across any BB with multiple successors, such as a loop exit node. So it prevents moving insertvalue
* across loop iterations
* to inner loop from outside

weiguozhi committed:
Fixing styles.
